webapp/src/ts/services/transitions/create-user-for-contacts.transition.ts

```diff
@@ -0,0 +1,173 @@
+import { ExtractLineageService } from '@mm-services/extract-lineage.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CreateUserForContactsTransition extends Transition {
+  constructor(
+    private dbService: DbService,
+    private createUserForContactsService: CreateUserForContactsService,
+    private extractLineageService:ExtractLineageService,
+  ) {
+    super();
+  }
+
+  readonly name = 'create_user_for_contacts';
+  private readonly REPLACE_PROPERTY = 'replace_forms';
+  private replaceForms;
+
+  /**
+   * @param {Object} settings - the CHT instance settings
+   * @return {Boolean} - whether the required config is present
+   */
+  init(settings) {
+    this.replaceForms = this.getReplaceForms(settings);
+    if (!this.replaceForms || !Array.isArray(this.replaceForms) || !this.replaceForms.length) {
+      console.warn(
+        `Configuration error. Config must define have a '${this.name}.${this.REPLACE_PROPERTY}' array defined.`
+      );
+
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @param {Array<Doc>} docs - docs to be saved
+   * @return {Boolean} - whether any of the docs from the batch should be processed
+   */
+  filter(docs: Doc[]) {
+    return !!docs.filter(doc => doc.type === 'data_record').length;
+  }
+
+  /**
+   * @param {Array<Doc>} docs - docs to run the transition over
+   * @returns {Promise<Array<Doc>>} - updated docs (may include additional docs)
+   */
+  async run(docs: Doc[]) {
+    const originalContact = await this.createUserForContactsService.getUserContact();
+    if (!originalContact) {
+      return docs;
+    }
+
+    if (this.createUserForContactsService.isReplaced(originalContact)) {
+      this.reparentReports(docs, originalContact);
+    }
+    const userReplaceDoc = this.getUserReplaceDoc(docs);
+    if (userReplaceDoc) {
+      return this.replaceUser(docs, userReplaceDoc, originalContact);
+    }
+
+    return docs;
+  }
+
+  private getTransitionConfig(settings) {
+    return settings[this.name];
+  }
+
+  private getReplaceForms(settings) {
+    return this.getTransitionConfig(settings)?.[this.REPLACE_PROPERTY] || [];
+  }
+
+  private getUserReplaceDoc(docs: Doc[]) {
+    return docs.find(doc => this.replaceForms.includes(doc.form)) as UserReplaceDoc;
+  }
+
+  private async replaceUser(docs: Doc[], userReplaceDoc: UserReplaceDoc, originalContact: Doc) {
+    const { replacement_contact_id } = userReplaceDoc.fields;
+    if (!replacement_contact_id) {
+      throw new Error('The form for replacing a user must include a replacement_contact_id field ' +
+        'containing the id of the new contact.');
+    }
+
+    const originalContactId = userReplaceDoc.contact._id;
+
+    const isOriginalContact = () => originalContact._id === originalContactId;
+    const isReplacedContact = () => this.createUserForContactsService.isReplaced(originalContact) &&
+      this.createUserForContactsService.getReplacedBy(originalContact) === originalContactId;
```

Comment:
It was hard to follow what was going on when I had them all combined on one line in the if-header.  But, I wanted to be able to short-circuit out and avoid running the logic in `isReplacedContact` if `isOriginalContact` returns `true`.  That is why I used functions instead of just assigning the values to variables...

Reply:
Thanks for the explanation. 
I understand the circuiting, but those two functions are in fact really cheap and just run some comparisons and nullchecks on some properties of the contact: 
```
  isReplaced(contact) {
    return !!contact.user_for_contact?.replaced;
  }

  getReplacedBy(contact): string {
    return contact.user_for_contact?.replaced?.by;
  }
```

I don't think there's a need to short-circuit. Could you please just make them constants?